### PR TITLE
HDFDataset / HDFDatasetWriter: Support for only using targets

### PR DESCRIPTION
**Pull request by patrick-wilken**
Necessary changes to support HDF files without 'inputs'.

Also, there was a check that targets have to be of equal sequence length in the HDFDatasetWriter. I tested targets with different lengths and I didn't find a reason why this check was there.

### CODE REVIEW

File: returnn/datasets/hdf.py [hunk 1]
```diff
@@ -270,14 +278,16 @@ def get_data(self, seq_idx, key):
     end_pos = self.file_seq_start[file_idx][real_file_seq_idx + 1]
 
     if key == "data":
+      assert 'inputs' in fin, "'data' key is reserved for 'inputs' in the HDF file, but 'inputs' not found in HDF file."
```
Comment (albertz) [inline returnn/datasets/hdf.py:281]: Maybe we might want to allow `"data"` as a possible entry in `targets/`?

File: returnn/datasets/hdf.py [hunk 2]
```diff
@@ -1287,27 +1297,35 @@ def dump_from_dataset(self, dataset, epoch=1, start_seq=0, end_seq=float("inf"),
...
+        else:
+          raise Exception("not sure which input data key to use from %r" % (data_input_keys,))
       else:
-        raise Exception("not sure which input data key to use from %r" % (data_input_keys,))
-    else:
-      default_data_input_key = data_input_keys[0]
+        default_data_input_key = data_input_keys[0]
+      progress_bar_data_key = default_data_input_key
     print("Using input data key:", default_data_input_key)
+    default_data_target_key = None
```
Comment (albertz) [inline returnn/datasets/hdf.py:1312]: Do we actually still need `default_data_target_key`?

**Comment by albertz** [comment]
Note that a HDF created with this new code might not be compatible with an old RETURNN (e.g. when "inputs" is missing). This is ok?

I think it's not necessary. You might also put just some dummy entry into "inputs", with 0 dim or so. Maybe the code is actually simpler then, as you don't need this more complex logic for `first_target_idx`, `num_input_keys`, etc.

**Review by JackTemaki — COMMENTED** [review]
The HDF using no targets will not work with the MetaDataset right now, as this will do calls like `dataset.get_seq_length(idx)["data"]` on the HDFDataset, which would now return an error.

I am working on the PR for the MetaDataset right now, then this should work. I think we should add test-cases for this, many dataset-related things have no coverage at all right now.

EDIT: Of course this error only happens in the case that the HDFDataset is used as "default dataset" inside the MetaDataset. Still, this should be supported.

**Comment by albertz** [comment]
@JackTemaki your comment is pretty independent from this PR here, right?

**Comment by JackTemaki** [comment]
> @JackTemaki your comment is pretty independent from this PR here, right? Then better let's discuss this elsewhere, and not mix up things, to keep the discussion focused here about this specific PR.

This was somewhat related to the question if we should allow the "data" key in the targets, and this is also something Patrick needs to consider when testing his code. I did not want to start a discussion about this issue here.

**Comment by albertz** [comment]
> > @JackTemaki your comment is pretty independent from this PR here, right?...
> 
> This was somewhat related to the question if we should allow the "data" key in the targets, ...

Why?

File: returnn/datasets/hdf.py (hunk 2 above)
Comment (patrick-wilken) [reply returnn/datasets/hdf.py:1312]: Only to create this `attr_numLabels = 'numLabels'` field in the HDF. I don't know if anyone uses it anymore though, it has been replaced by `'targets/size'`, see line 163.

File: returnn/datasets/hdf.py (hunk 1 above)
Comment (patrick-wilken) [reply returnn/datasets/hdf.py:281]: Yes, would probably be good. So just
```
if key == "data" and 'targets/data/data' not in fin:
```
here?

To actually create an HDF with data as target we would need to change the HDFDatasetWriter.

File: returnn/datasets/hdf.py (hunk 2 above)
Comment (albertz) [reply returnn/datasets/hdf.py:1312]: Yea, since a long time. I think it would be safe to remove.

File: returnn/datasets/hdf.py (hunk 1 above)
Comment (albertz) [reply returnn/datasets/hdf.py:281]: Yea. I'm not sure we need that logic now, but we should keep that in mind.

**Comment by patrick-wilken** [comment]
>The HDF using no targets will not work with the MetaDataset right now.

True, I fixed the same thing for CombinedDataset by having `default_data_key` as a parameter of the dataset. Maybe we can think of something more general.

>Note that a HDF created with this new code might not be compatible with an old RETURNN (e.g. when "inputs" is missing). This is ok?

For me, yes 😄 

>I think it's not necessary. You might also put just some dummy entry into "inputs", with 0 dim or so. Maybe the code is actually simpler then, as you don't need this more complex logic for first_target_idx, num_input_keys, etc.

Hmm, you would also need the sequence lengths for the "data" key. So store an additional 0 for each sequence. I would rather keep the HDF file clean than avoiding this logic in the code.

File: returnn/datasets/hdf.py (hunk 2 above)
Comment (patrick-wilken) [reply returnn/datasets/hdf.py:1312]: What about `inputPattSize`? Shall I remove both in one go?

**Comment by JackTemaki** [comment]
> > The HDF using no targets will not work with the MetaDataset right now.
> 
> True, I fixed the same thing for CombinedDataset by having `default_data_key` as a parameter of the dataset. Maybe we can think of something more general.

Okay, then this is identical to how I planned to solve this (giving this parameter to the "Dataset")

File: returnn/datasets/hdf.py (hunk 2 above)
Comment (patrick-wilken) [reply returnn/datasets/hdf.py:1312]: Ah no, for sparse "data" it seems to be used.

**Review by albertz — APPROVED** [review]

**Comment by albertz** [comment]
What's the state here? @JackTemaki ? Ready to merge?

**Review by JackTemaki — APPROVED** [review]

**Comment by JackTemaki** [comment]
> What's the state here? @JackTemaki ? Ready to merge?

I thought there would be still some open discussion, but for me it is fine.

**Pull request merged**